### ISSUE TITLE
windows_feature (and other not working with test kitchen base box 

### ISSUE THREAD

**Issue by jmahowald** [test-kitchen/test-kitchen#655]
both for using windows features generally, and using the sql server cookbook specifically I run into issues when I'm executing via kitchen converge but when I execute on the machine via the command prompt  
i.e.
chef-client -c $env:TEMP\kitchen\client.rb -j $env:TEMP\kitchen\dna.json -z

it works.  I'm not sure if this is a base box issue, or an issue with policy, etc.

i.e.

```
[2015-03-26T09:40:04-07:00] ERROR: Converge failed with error message windows_feature[TelnetClient] (win2012::default line 7) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0, 42, 127, 3010], but received '-2146498298'
      ---- Begin output of C:\Windows\sysnative\dism.exe /online /enable-feature /featurename:TelnetClient /norestart  ----
      STDOUT: Deployment Image Servicing and Management tool
      Version: 6.3.9600.17031
       
      Image Version: 6.3.9600.17031
       
      Enabling feature(s)
       
      Error: 0x800f0906
       
      The source files could not be downloaded. 
      Use the "source" option to specify the location of the files that are required to restore the feature. For more information on specifying a source location, see http://go.microsoft.com/fwlink/?LinkId=243077.
```

**Comment by jmahowald** [comment]
doh, forgot to post the box:
mwrock/Windows2012R2

**Comment by fnichol** [comment]
@jmahowald was that output from the chef-client run itself? Would it be possible to gist more of the kitchen run output (assuming there aren't secrets/credentials/etc)?

Adding @test-kitchen/windows to see if we can help

**Comment by jdmundrawala** [comment]
i bet running through task scheduler would fix this :trollface:

**Comment by fnichol** [comment]
I was wondering…this was a simplifying assumption we made with running chef-{solo,client} in Test Kitchen instances (running directly as the logged in user). There might be some work to do around this but I'm not sure what that would be at present.

**Comment by jdmundrawala** [comment]
yea, an async workflow might be better. I've argued we should do this for bootstrapping windows nodes because it gets around the whole winrm timeout mess. That being said, we did this to get around the kb of death in the knife bootstrap code.

**Comment by jmahowald** [comment]
Having vague flashbacks to funny policy settings and execution policies with powershell when I was playing around with knife windows

actually, that's a thought. Once my stupid box is back up I'll test with knife winrm somehow

**Comment by adamedx** [comment]
In general I'd expect kitchen with the winrm transport to face the same issues as knife-windows. If ktichen is somehow less functional, there is definitely something to investigate.

**Comment by smurawski** [comment]
First off.. why the heck is Telnet getting installed? :)

@jmahowald  Is this a server core box?  It looks like the error is related to failing to find the source to install the feature.  Server Core has a number of feature marked "removed" where the bits to enable that feature aren't on the box.  You need to point DISM to a windows image with the bits available.  If the cookbook doesn't support this, it may be an issue to file against it.

There can be a WinRM issue with the source you point to, if the computer account doesn't have rights to the filesystem where the image is (since CredSSP or Kerberos delegation is likely not configured).

**Comment by cmluciano** [comment]
I'm with @smurawski. The source error is something that I saw recently when trying to "slim down" a box and I needed to point to a disk for a specific feature.

**Comment by jmahowald** [comment]
Tried with other features, like "MicrosoftWindowsPowerShellISE" I'd rather not go into the inanity of why telnet was being used.  Also notice with using sql_server cookbook.   And more specifically commands that are invoked directly on the windows CLI aren't working when being invoked in the winrm transport, as used by test-kitchen I believe (or on some execution policy in the base box)

**Comment by smurawski** [comment]
@jmahowald Is this a server core box?  That'll be an issue for a lot of roles and features.  Like, you can't install the ISE without some of the UI features  and the ISE bits are also likely "removed".  I don't have a core box handy (I'm on the road), but that's likely the source of the DISM woes.

And yes, some commands operate differently in the winrm context, especially due to credential delegation, but I can't respond to how something behaves without an example.

**Comment by mwrock** [comment]
so the base box i typically use mwrock\Windows2012R2 https://atlas.hashicorp.com/mwrock/boxes/Windows2012R2 has most features uninstalled. I take advantage of "Features on Demand" in the base box prep to eliminate about 2gb in the .box file. The unfortunate side effect is that there are some features that can not be reinstalled over simple windows update and it needs the actual installation media to install that feature. Examples are telnet and some of the AD features.

Coincidentally I am working on a blog post that demonstrates using test kitchen to test multi node windows deployments using a primary and backup AD controller as an example and thus need a different base box. Just today I uploaded the Hyper-V provider : https://atlas.hashicorp.com/mwrock/boxes/Windows2012R2Full and will be adding VirtualBox tonight.

I prepped this box the same way (see http://www.hurryupandwait.io/blog/in-search-of-a-light-weight-windows-vagrant-box) except left all features uninstalled and also sysprepped the box. Its a 5gb box.

So the bottom line is that a chef run will hit this issue on SOME but not all features that have been previously removed.

**Comment by mwrock** [comment]
HAH! just saw you are using my box so this is definitely the issue :)

**Comment by jmahowald** [comment]
Nice to hear from you @mwrock - I've been following your posts, and made this issue as bait to hope to lure you out:)

The thing is, those commands work fine when run via command prompt, i.e. last night I was able to get

```
 dism /online /enable-feature /featurename:TelnetClient 
```
when typed into a prompt.  Or for SQL server, 

start "" /wait "C:\Users\ADMINI~1\AppData\Local\Temp\kitchen\cache\SQLEXPR_x64_ENU.exe" /q /ConfigurationFile=C:\Users\ADMINI~1\AppData\Local\Temp\kitchen\cache\ConfigurationFile.ini

would work when run on CLI but not via kitchen converge.


then I did some more tests this morning and believe that running chef client directly on a windows prompt works.

I'm going to do some more triangulation, i.e. I'll eliminate test kitchen by using a chef server provisioner and use knife windows to try as well.  I'm also going to rerun all my failed attempts last night in a chef run directly on the machine.

a) base box (I have an AMI I can uuse)
b) winrm/powershell as used by kitchen  vagrant and windows cookbooks.

**Comment by mwrock** [comment]
One thing you could try to unblock a scenario like this is to download the eval win2012R2 from tecthet: http://www.microsoft.com/en-us/evalcenter, mount the iso to the VM and then run:

```
C:\Windows\sysnative\dism.exe /online /enable-feature /featurename:TelnetClient /norestart /source:d:\sources
```

This assumes the iso is mounted to D:\

Its possible this might NOT work if the latest iso has diverged much from the eval iso I used to create the box.

**Comment by jmahowald** [comment]
btw, on that base box, I couldn't get RDP working until I added this to my run

```
windows_registry 'HKLM\SYSTEM\CurrentControlSet\Control\Terminal Server\WinStations\RDP-Tcp' do
  values 'UserAuthentication' => 0, 'SecurityLayer' => 1
end

```

**Comment by smurawski** [comment]
@jmahowald The reason the DISM command is working from the command line but not via WinRM is the logged on user context has access to call into the Windows Updates API to retrieve feature details, while WinRM which operates in a non-interactive context and cannot use that API.  Matt's solution is the workaround for working in the winrm context.

**Comment by mwrock** [comment]
Oh maybe I'm wrong about the TelnetClient. So if it does try to get it from windows update, test kitchen will fail because it is driving the chef run over winrm. Windows update will block all requests over winrm and the only work around i am aware of is to go local or to run via a scheduled task.

**Comment by smurawski** [comment]
Regarding the start command to install SQL, that may be a problem.  I have installed SQL Server over WinRM (I ran my SQL boxes as server core at Stack Exchange and we did all the sql installs over winrm), but I didn't do it via the start command.

**Comment by jmahowald** [comment]
I went to sleep last night having nightmares of the rabbit hole I was about to go down after ending up at this post on credssp :)  (https://social.technet.microsoft.com/forums/windowsserver/en-US/ab722e45-8fca-40bc-acfb-499c11569411/enabling-and-using-credssp)

**Comment by smurawski** [comment]
Yeah, CredSSP or Kerberos delegation might help or if the process was kicked off by the task scheduler, which operates in an interactive context.  Just FYI, after ChefConf next week, I'm starting an overhaul of the SQL Server cookbook, so I'll be bumping into this pain point shortly and working on a solution.

**Comment by mwrock** [comment]
I've seen similar issues with SQL. depending on your setup, it will go out through windows update to download some bits. This has been hit and miss for me though it has been a while since i was installing sql regularly. Could be if it also needs to install .net 3.5 in the same go. This also tends to break because it often needs to be rebooted at least once.

**Comment by smurawski** [comment]
Yeah - .net 3.5 and disabling the sql check for updates are key in installing over winrm.

**Comment by jmahowald** [comment]
Didn't have issues with .net 3.5 once I switched to use 2012.  pushed to private repo an hour ago change to turn off updates in the sql server cookbook template (there's a process I need to go through to push for my company, and was waiting till I got everything working).

**Comment by mwrock** [comment]
Here is an example that I use when installing OctopusDeploy that dies in winrm but succeeds if you wrap the calls in a task:

```
batch_path = "#{node["platform_octopus"]["home_path"]}/octopus.bat"
template batch_path do
  source "octopus.bat.erb"
  action :create
end

windows_task 'octopus' do
  user "blah"
  password "blahblahblabadyblah"
  cwd "c:\\Program Files\\Octopus Deploy\\Tentacle"
  command batch_path
  frequency :monthly
  frequency_modifier 12
  run_level :highest
  action :create
  not_if  { ::File.exists?("#{node['platform_octopus']['home_path']}\\Tentacle\\Tentacle.config") }
end

windows_task 'octopus' do
  action :run
  not_if  { ::File.exists?("#{node['platform_octopus']['home_path']}\\Tentacle\\Tentacle.config") }
end
```

The template `octopus.bat.erb` has the command calls that would typically fail in winrm. The trick is that firing the task is a "fire and forget" operation. So there needs to be something you can poll for to tell if things are done. In the case of OctopusDeploy this looks like:

```
      uri = URI.parse("http://#{node['platform_octopus']['server_uri']}/api/machines/discover?host=#{node['ipaddress']}&port=#{node["platform_octopus"]["tentacle_port"]}")
      Chef::Log.info("Octopus API Call: #{uri}")

      req = Net::HTTP::Get.new(uri.request_uri)
      req['X-Octopus-ApiKey'] = node["platform_octopus"]["apikey"]
      status = [0]
      res = nil

      # The task above will fire asynchronously so we wait
      Timeout::timeout(60) do
        until ["200"].include?(status)
          sleep 5

          res = Net::HTTP.start(uri.host, uri.port) do |http|
            http.request(req)
          end

          status = res.code
        end
      end

      unless(["200"].include?(status))
        throw "Octopus request failed. Code: #{res.code}. Body: #{res.body}"
      end
```

**Comment by jmahowald** [comment]
In what I'm sure is not surprising to @smurawski 
I setup my local node as chef client as well with knife windows bootstrap and confirmed that the issue isn't test kitchen as much as winrm.  

So I have recipes that install features, sql server, that work fine when executed on the box itself with chef-client run via windows CLI, but fail with winrm, both via test-kitchen and  knife winrm.

**Comment by mwrock** [comment]
i think we need a telnet transport :)

**Comment by jmahowald** [comment]
well I still want to confirm with another box I had setup as an AMI, which I don't think I've seen anything like this on, that way can triangulate if it's a box configuration.  For work around, I'm thinking of making my chef run just create a scheduled task and build up the real run list and attributes in there.  Question will be reading output for debugging purposes.  I didn't have a lot of success with the event viewer handler last I tried but that was a while ago.

**Comment by jmahowald** [comment]
This seems pretty related

http://serverfault.com/questions/493170/why-can-a-domain-admin-run-a-powershell-cmd-locally-but-connecting-over-winrm-w


which leads to:

http://blogs.technet.com/b/heyscriptingguy/archive/2012/11/14/enable-powershell-quot-second-hop-quot-functionality-with-credssp.aspx

but it's beyond my current capabilities to figure out how to map this into the problem at hand.

**Comment by smurawski** [comment]
@jmahowald I'm on the road for the next week or so, but I'll try to dig into this as time allows.  Otherwise, it'll be top of my list after ChefConf.

**Comment by mwrock** [comment]
Here's my 5 cents:

I've struggled quite a bit with this over the past few years and tried to capture much of my conclusions here: http://www.hurryupandwait.io/blog/safely-running-windows-automation-operations-that-typically-fail-over-winrm-or-powershell-remoting

In your case you are trying to install a feature that was previously removed via `uninstall-windowsfeature` explained in detail here: https://technet.microsoft.com/en-us/library/jj127275.aspx. Windows will attempt to get the feature from windows update if not given an explicit source. If it cannot get the feature from windows update, as is the case here, it will ask for a source. Since you are remoted in via winrm, windows will always fail unless you sneak in via a remote task. This is a separate issue from CredSSP as the 2nd blog link above points too. That comes into play with "2nd hop" authentication. For example, trying to access a network share from your winrm session. Its a separate issue from windows updates but it is certainly a cousin in the sense that it is a common remoting complication. By default, CredSSP is not enabled and your security token in such cases cannot be delegated beyond your current remote session.

The fact that you lose visibility inside a scheduled task is a painful downside to that approach but there are tricks there but nothing that I'm aware of that chef does for you. I have a side project, boxstarter.org,  where I wrap chocolatey and intercept, MSIEXEC, windows update calls and windows feature installs. I funnel them through a scheduled task and capture stdout and stderr of the tasks process and forward them to the console and a log as well as monitor the task and all child processes for a hung state. The source code is here: https://github.com/mwrock/boxstarter/blob/master/BoxStarter.Common/Invoke-FromTask.ps1.

Its a real shame that the MS remoting solutions get you 95% of the way there. You can always travel the remaining 5% as long as you don't mind passing through rings of fire. I'm confident it will get better....for my children.

**Comment by carpnick** [comment]
We have a design that allows WinrRM but gets around the context issue and allows redirection of chef output.  It is a feature we plan on adding here once the transport layer is officially released.

During bootstrap we create a Windows Service that has a web service in it.  We also deploy a web service client to the windows service above.  We then start chef in the Windows Service and redirect the output over the local web client and print it to screen.

We run the entire thing over WinRM and get rid of all memory and context issues.  WinRM 1.3 has the timeout issues fixed, so you can specify the timeout, knife-windows supports this in [master] (https://github.com/chef/knife-windows/pull/186)

A crude picture of the architecture that works very well for us:
![image](https://cloud.githubusercontent.com/assets/6966179/6873163/c54fb87a-d484-11e4-8196-211ee2ebbe22.png)

**Comment by jmahowald** [comment]
Thanks all for your time.  I don't have a fleet of machines to worry about at this point in time,and I have a much better understanding of the problem and what the future options are.

I'll be at Chefconf for parts of next week, so hopefully I get to bump into some of you at some point.

(Is there a punt option on issues :)

**Comment by jmahowald** [comment]
In case anyone else hits this link before Nick's more full solution is out, my workaround was to bookend scheduled task creation and deletion

At the beginning of my run list I have
```
windows_task 'chef_kitchen_converge' do
  user 'vagrant'
  password 'vagrant'
  command 'C:\chef\converge.bat' 
  run_level :highest
  frequency :minute
  frequency_modifier 10
end
```

and then at the very end of the run list I remove it

```

windows_task 'chef_kitchen_converge' do
 action :delete
end
```
Since the failures come in the intervening recipes, I'll have the scheduled task happen only when the run didn't complete successfully, which is my first run of converge.

**Comment by jmahowald** [comment]
back on this issue.  Wanted to clarify with @carpnick if he was talking about a purely internal solution for his company or if there were plans to release that

**Comment by carpnick** [comment]
@jmahowald, There are 2 pieces above that I mentioned.
* The WinRM workaround design - Using a Windows service.  That is currently in-house, and if you feel that would add value, then we can definitely open source.
* Test-Kitchen changes to call custom chef-client call, handle reboot support, custom bootstrap - We have not investigated test-kitchen quite yet on the Windows side.  It is being prioritized internally a bit lower as we try to get our first application to prod.  But we plan on revisiting this one, and it will definitely be open source when we get to it.

**Comment by DaveCasey** [comment]
If we start with a Windows2008 box with CredSSP enabled settings baked in, is there something that can be done on the Test Kitchen side to get Visual Studio/SQL Server/etc cookbooks to run?

**Comment by mwrock** [comment]
I'm assuming you are installing these from a network share and not a public HTTP endpoint? That would certainly require 2nd hop authentication to the share. Unfortunately even with CredSSP enabled at both ends, the ruby WinRM gem cannot use the CredSSP protocol.

I've handled this with scheduled tasks in the pass which is messy to say the least. I've been thinking that if you are going from a windows host to a windows test instance (seems like most, including most of my use cases, come from a linux host), it would be interesting to come up with a new `Transport` that levereges native Powershell remoting instead of the ruby WinRM gem.

Such a transport could authenticate via CredSSP. However, as far as I know no one has developed this.

**Comment by jmahowald** [comment]
@carpnick I think that windows service would be of great general use.  We've gone with the windows task route, along with a loop and guard to wait until it finishes, and it feels like we're close to having an LWRP route.

I know that @adamedx strongly cautioned against using CredSSP which made me happy as my readings into that made my brain bleed.

I'm very curious with the recently announced support of [ssh on windows](http://www.theregister.co.uk/2015/06/02/openssh_windows/) if that'll have the same limitations on the update API as the winrm route.

**Comment by mwrock** [comment]
Taking the scheduled task route,  have you considered redirecting standard output/error of the powershell process running inside the task? Then locate the process, grab the stream and route it back to the console? This could make the scheduled task "appear" interactive.

I've done something similar with powershell here: https://github.com/mwrock/boxstarter/blob/master/BoxStarter.Common/Invoke-FromTask.ps1

This redirects the streams to a file and then uses a `FileStream` to stream to the file to the console until the process terminates. It adds hang detection too in the event a dialog is invoked and the task then never returns.

Using the FileStream is a bit awkward and could be improved upon by avoiding the file and just referencing the stream exposed by `Process.StandardOutput` I've often thought of converting this cmdlet to a gem but alas...thought does not emit time. 

Anyways, the thing I hate about scheduled tasks (well one thing) is its fire and forget nature. If you could stream realtime output, that would be way cool. Doesn't matter so much for regularly converging infrastructure but it is invaluable for dev/test and prod debugging.

**Comment by dpiessens** [comment]
Could someone describe if/how you're laying out test kitchen to do this on Windows? I'm running into the same thing trying to use Powershell DSC, test-kitchen and vagrant together. My converges work fine on a live chef node but fail with a test kitchen run when it hits a windows_feature resource or bootstrapping PowerShell5. 

I'd also be happy to contribute if someone could point me at details on how chef compiles and uploads the resource so we could abstract the run into a task.

**Comment by smurawski** [comment]
Yeah, you can't install features that need to contact windows update or install MSU's over winrm.  I've got a start on a chef-zero provisioner that uses a scheduled task to run, but it is very early on. 

https://github.com/smurawski/chef-zero-scheduled-task

It should work with just as a direct swap out for chef-zero as a provisioner.  As it stands, it'll start a scheduled task to run chef-client and then poll until the task finishes and dump the log.  I'll be continuing to improve that and eventually add it into core test-kitchen.

**Comment by carpnick** [comment]
@jmahowald and whoever else is interested in the winrm workaround.
https://github.com/ebsco/chefservice

This client/web service/ windows service solution will unblock using WinRM natively for chef-client runs.  The only difference is not calling chef-client via the knife winrm command.

Feel free to give any feedback here or in the repo as this has been working quite well internally for us.


Next step is to integrate into Kitchen

**Comment by sneal** [comment]
Vagrant [wraps all WinRM commands in a scheduled task](https://github.com/mitchellh/vagrant/blob/master/plugins/communicators/winrm/communicator.rb#L185-L214), including Chef runs. I'm sure this saved me from hundreds of 'this works when I run it locally on the box, but not via Vagrant' type of questions. Some of my thoughts on this:

1. Make the scheduled task wrapper optional, perhaps used by default in test-kitchen
2. knife support for using a scheduled task to execute Chef.
3. Share the components that wrap winrm commands in a scheduled task in a reusable gem used by test-kitchen, vagrant, etc.
4. Share code for file transfers, via winrm-fs or something.

**Comment by mwrock** [comment]
what @sneal said :+1: 

STDD (Scheduled Task Driven Development) - Its the future...

**Comment by cheeseplus** [comment]
This meandered a bit off topic and it seems there isn't in fact an issue but perhaps a feature request. Closing.
